Bug22373: Enhance documentation for RangeIndex

mikapfl opened the pull request:
Corrects documentation for RangeIndex and enhances documentation for RangeIndex and RangeIndex.from_range
closes #22373
passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

mikapfl committed:
core/indexes/range.py: Add dtype documentation to RangeIndex.

mikapfl committed:
core/indexes/range.py: Correct order of arguments in RangeIndex docstring. Fixes #22373

mikapfl committed:
core/indexes/range.py: Add fastpath argument to the RangeIndex docstring.

mikapfl committed:
core/indexes/range.py: Clean up formatting of RangeIndex docstring according to pandas documentation styleguide.

mikapfl committed:
core/indexes/range.py: Full docstring for RangeIndex.from_range.

pep8speaks commented:
Hello @mikapfl! Thanks for submitting the PR.

 - In the file [`pandas/core/indexes/range.py`](https://github.com/pandas-dev/pandas/blob/201408e2ecd9706fc215e12f780a81c45c9c5aa0/pandas/core/indexes/range.py), following are the PEP8 issues :

> [Line 43:80](https://github.com/pandas-dev/pandas/blob/201408e2ecd9706fc215e12f780a81c45c9c5aa0/pandas/core/indexes/range.py#L43): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 47:80](https://github.com/pandas-dev/pandas/blob/201408e2ecd9706fc215e12f780a81c45c9c5aa0/pandas/core/indexes/range.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 137:80](https://github.com/pandas-dev/pandas/blob/201408e2ecd9706fc215e12f780a81c45c9c5aa0/pandas/core/indexes/range.py#L137): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

mikapfl committed:
pandas/core/indexes/range.py: Make pep8speaks happy.

datapythonista commented on pandas/core/indexes/range.py:
```diff
@@ -30,23 +30,30 @@ class RangeIndex(Int64Index):
+        If of type `int` and `stop` is not given, interpreted as `stop`
+        instead.
+    stop : int, default 0
+    step : int, default 1
```
It'f be good to have these documented. Also, `start`, which would be good to have a description, besides the behavior described.

mikapfl replied:
Re documenting `start`, `stop` and `step`: I also thought that it would be nice to document them, and figured we could just recycle the [python documentation for `range`](https://docs.python.org/3/library/stdtypes.html#typesseq-range). However, the documentation for the arguments is very basic in python as well: "`stop`: The value of the stop parameter". Maybe there is not much else to say?

But I'll try to adapt the rest of the python `range` docs (i.e. the long description) and try to come up with some useful examples.

datapythonista replied:
I think it can be useful to say whether the `stop` value is included or not. It's not obvious to me that `range(0, 10)` doesn't include the `10`.

Also, for beginner users, something as simple as  `start : The first value of the generated range.`  can be useful. `step` can be tricky, do we support negative values, as `range`?

I do think there is a lot to say. :)

jreback replied:
> think it can be useful to say whether the stop value is included or not. It's not obvious to me that range(0, 10) doesn't include the 10.

this is standard python behavior to NOT include the rhs on ranges.

datapythonista reviewed (commented):
Thanks for the changes, looking good. If you want to add examples as part of this PR, that would be great (we can do it in a separate PR too).

Also, if you can add periods at the end of the `See Also` items description, that would be great.

jreback commented on pandas/core/indexes/range.py:
```diff
@@ -30,23 +30,30 @@ class RangeIndex(Int64Index):
     copy : bool, default False
         Unused, accepted for homogeneity with other index types.
+    name : object, optional
+        Name to be stored in the index
```
this constructor should be discouraged for using integers on ranges, this is the reason we have ``.from_range``. Actually folks should generally NOT use specific constructors anyhow. ``pd.Index`` dispatches in most cases. Only if you have very very specific requreiments do you need to use this constructor.

datapythonista commented:
Agree that not including the stop of the range is standard, my point is that it's not obvious unless you know about this standard. So, I think it's worth mentioning it in the documentation.

If using the constructor is discouraged, we should also say it in the description I guess.

mikapfl committed:
RangeIndex docstring: add explanation and examples for RangeIndex ranges.
Explanation is based on the python documentation for ranges.

mikapfl committed:
RangeIndex docstring: document that users should prefer to pass a range to Index.
Index docstring: add example of passing a range to construct a RangeIndex.

mikapfl commented:
@datapythonista I have added documentation and examples showing the half-inclusive behaviour of RangeIndex. The documentation is based on the python range documentation.

@jreback I have added langeuage and examples to Index and RangeIndex to guide users towards using Index(range(…)) instead of RangeIndex(…).

datapythonista reviewed (commented):
Can you change the default of `tupleize_cols` from `(default: True)` to `default True

And finish the comment in `Notes` with a period.

alimcmaster1 commented:
@mikapfl could you rebase and merge master if you are still looking to take this on? If not I can potentially take a look

datapythonista commented:
@alimcmaster1, if you want to open a new PR with this. Closing as this seems abandoned.
